File: app/src/main/java/limor/tal/mytodo/Task.java
package limor.tal.mytodo;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

@Entity(tableName = "tasks")
public class Task {
    @PrimaryKey(autoGenerate = true)
    public int id;

    public String description;
    public Long dueDate;
    public Long dueTime; // Milliseconds since midnight
    public String dayOfWeek;
    public boolean isRecurring;
    public String recurrenceType;
    public boolean isCompleted;
    public int priority;
    public Long completionDate;
    public Integer reminderOffset; // Minutes before due time: null (no reminder), 0 (at time), 15, 30, or 60
    public String reminderDays; // For daily recurring tasks: null (all days), or comma-separated day indices (0=Sunday, 1=Monday, etc.)
    public Integer manualPosition; // null = automatic ordering, integer = manual position from drag operations
    public String firestoreDocumentId; // Firestore document ID for cloud sync
    public Long createdAt; // Timestamp when task was created
    public Long updatedAt; // Timestamp when task was last updated
    public Long deletedAt; // Timestamp when task was deleted (null = not deleted)
    
    // FamilySync integration fields
    public String sourceApp; // "familysync" if imported from FamilySync
    public String sourceTaskId; // Original FamilySync task ID
    public String sourceGroupId; // FamilySync group ID
    public String familySyncAssigneeId; // FamilySync assignee ID
    public String familySyncCreatorId; // FamilySync creator ID

    public Task(String description, Long dueDate, String dayOfWeek, boolean isRecurring, String recurrenceType, boolean isCompleted, int priority) {
        this.description = description;
        this.dueDate = dueDate;
        this.dueTime = null;
        this.dayOfWeek = dayOfWeek;
        this.isRecurring = isRecurring;
        this.recurrenceType = recurrenceType;
        this.isCompleted = isCompleted;
        this.priority = priority;
        this.completionDate = null;
        this.reminderOffset = null; // No reminder by default
        this.reminderDays = null; // All days by default
        this.manualPosition = null; // Automatic ordering by default
        this.createdAt = System.currentTimeMillis();
        this.updatedAt = System.currentTimeMillis();
    }

    /**
     * Check if this task is imported from FamilySync
     */
    public boolean isExportedFromFamilySync() {
        return "familysync".equals(sourceApp) && sourceTaskId != null;
    }

    @Override
    public String toString() {
        return "Task{" +
                "id=" + id +
                ", description='" + description + '\'' +
                ", dueDate=" + dueDate +
                ", dueTime=" + dueTime +
                ", dayOfWeek='" + dayOfWeek + '\'' +
                ", isRecurring=" + isRecurring +
                ", recurrenceType='" + recurrenceType + '\'' +
                ", isCompleted=" + isCompleted +
                ", priority=" + priority +
                ", completionDate=" + completionDate +
                ", reminderOffset=" + reminderOffset +
                ", reminderDays='" + reminderDays + '\'' +
                ", manualPosition=" + manualPosition +
                '}';
    }
}

File: web/mytodo-web/src/models/Task.js
// Task model that matches the Android app's Task class
import { TaskConstants } from '../constants/TaskConstants';

export class Task {
  constructor(data = {}) {
    this.id = data.id || null;
    this.description = data.description || '';
    this.dueDate = data.dueDate || null; // Timestamp
    this.dueTime = data.dueTime || null; // Milliseconds since midnight
    this.dayOfWeek = data.dayOfWeek || TaskConstants.DAY_NONE;
    this.isRecurring = data.isRecurring || false;
    this.recurrenceType = data.recurrenceType || null;
    this.isCompleted = data.isCompleted || false;
    this.priority = data.priority || 0;
    this.completionDate = data.completionDate || null;
    this.reminderOffset = data.reminderOffset !== undefined ? data.reminderOffset : null; // Minutes before due time
    this.reminderDays = data.reminderDays || null; // For daily recurring tasks
    this.manualPosition = data.manualPosition || null;
    this.createdAt = data.createdAt || Date.now();
    this.updatedAt = data.updatedAt || Date.now();
    this.deletedAt = data.deletedAt || null;
    
    // FamilySync integration fields
    this.sourceApp = data.sourceApp || null;
    this.sourceTaskId = data.sourceTaskId || null;
    this.sourceGroupId = data.sourceGroupId || null;
    this.familySyncAssigneeId = data.familySyncAssigneeId || null;
    this.familySyncCreatorId = data.familySyncCreatorId || null;
  }

  // Convert to Firestore format
  toFirestore() {
    const firestoreData = {
      description: this.description,
      dueDate: this.dueDate,
      dueTime: this.dueTime,
      dayOfWeek: this.dayOfWeek,
      isRecurring: this.isRecurring,
      recurrenceType: this.recurrenceType,
      isCompleted: this.isCompleted,
      priority: this.priority,
      completionDate: this.completionDate,
      reminderOffset: this.reminderOffset,
      reminderDays: this.reminderDays,
      manualPosition: this.manualPosition,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      deletedAt: this.deletedAt
    };

    // Include FamilySync fields if they exist
    if (this.sourceApp) firestoreData.sourceApp = this.sourceApp;
    if (this.sourceTaskId) firestoreData.sourceTaskId = this.sourceTaskId;
    if (this.sourceGroupId) firestoreData.sourceGroupId = this.sourceGroupId;
    if (this.familySyncAssigneeId) firestoreData.familySyncAssigneeId = this.familySyncAssigneeId;
    if (this.familySyncCreatorId) firestoreData.familySyncCreatorId = this.familySyncCreatorId;

    return firestoreData;
  }

  // Create from Firestore data
  static fromFirestore(id, data) {
    const task = new Task(data);
    task.id = id;
    return task;
  }

  // Format due date for display
  getFormattedDueDate() {
    if (!this.dueDate) return null;
    const date = new Date(this.dueDate);
    return date.toLocaleDateString();
  }

  // Format due time for display
  getFormattedDueTime() {
    if (!this.dueTime) return null;
    const hours = Math.floor(this.dueTime / (1000 * 60 * 60));
    const minutes = Math.floor((this.dueTime % (1000 * 60 * 60)) / (1000 * 60));
    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`;
  }

  // Get priority color
  getPriorityColor() {
    switch (this.priority) {
      case 3: return '#f44336'; // High - Red
      case 2: return '#ff9800'; // Medium - Orange
      case 1: return '#4caf50'; // Low - Green
      default: return '#9e9e9e'; // Default - Grey
    }
  }

  // Get priority text
  getPriorityText() {
    switch (this.priority) {
      case 3: return 'High';
      case 2: return 'Medium';
      case 1: return 'Low';
      default: return 'Normal';
    }
  }

  // Check if task is overdue
  isOverdue() {
    if (!this.dueDate || this.isCompleted) return false;
    const now = new Date();
    const dueDate = new Date(this.dueDate);
    return dueDate < now;
  }

  // Check if task is due today
  isDueToday() {
    if (!this.dueDate) return false;
    const today = new Date();
    const dueDate = new Date(this.dueDate);
    return today.toDateString() === dueDate.toDateString();
  }

  // Check if task is imported from FamilySync
  isFromFamilySync() {
    return this.sourceApp === 'familysync' && this.sourceTaskId;
  }

  // Check if task is deleted
  isDeleted() {
    return this.deletedAt !== null && this.deletedAt !== undefined;
  }
}
